### DISCUSSION TITLE
Support the Rego language in Code Blocks

### DISCUSSION THREAD

**Discussion by crqrdotcom** [outline/outline#4764] [Ideas]
Follow-up PR based on this discussion: https://github.com/outline/outline/pull/4771

--- 

Original message:

Hey 👋🏽 

I'd like to add support for the Open Policy Agent's [Rego language](https://www.openpolicyagent.org/docs/latest/policy-language/) in Code Blocks. It's supported by both Prism.js and Refractor.

I've done some local testing and seems like the following changes need to be done:

- Add an entry to `LANGUAGES` in `shared/editor/plugins/Prism.ts`
- Import the `refractor/lang/rego` package and register it in `shared/editor/nodes/CodeFence.ts`

Additionally - if this is ultimately a requirement - the `@types/refractor` package has to be updated (it's [missing the types definition for `refractor/lang/rego`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/refractor/lang)).

I've created a branch in [wtfiscrq/outline](https://github.com/wtfiscrq/outline) with the relevant changes: [main...wtfiscrq:outline:feat/rego-lang-support](https://github.com/outline/outline/compare/main...wtfiscrq:outline:feat/rego-lang-support).

Can I open a PR with this changes?

P.S.: _I'm also ok with updating [@types/refractor](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/refractor) although it'll probably take a long time to get the PR approved there (never contributed to that project before)._

**Comment by tommoor** [comment]
Sounds reasonable although based on the lack of types I'd suspect this is a very rarely used language. The problem right now is that all language definitions are loaded with the editor, so it adds weight to every single page load. It would be great to figure out how to load these dynamically so that we could essentially add all languages, even the rare ones.

**Reply by CHUYKOPS** [reply]
Hey. Will be better if admin can define languages in settings? It will speed up page loading

**Reply by tommoor** [reply]
I'm not sure the admin is in a position to make this decision for all users on the team. The solution for both would end up very similar though as there is not really any technical difference between loading just the language in the current code block and loading just the languages an admin has chosen.

**Reply by crqrdotcom** [reply]
Thanks for the quick reply! Rego is quite used but it's more a niche language (used to write policies) that might not be of most's interest.

I took an attempt at the _dynamically loading_ idea. I'm not very versed in TS/Webpack but I figured that using [dynamic imports](https://webpack.js.org/guides/code-splitting/#dynamic-imports) could be one/the solution for this problem?

Take a look at the latest commit to my branch: https://github.com/outline/outline/compare/main...wtfiscrq:outline:feat/rego-lang-support#diff-2738f362dd897eb6ddb8d04bfc5e27fb91e6e60213ad14420a6be8515f42ef7eR101-R110

---

With those changes, I was able to go to the editor with some code blocks that were properly rendered almost immediately after page load.

If you toggle to other languages (with the network tab open), you'll see it loading the chunk that contain the language you just selected. It'll only do one load per language so if you add another block using the same language you'll notice it won't trigger a new request.

I did these tests with `NODE_ENV=development` on a Chromium-based browser. I'm not 100% sure if this may have other implications in production or in other browsers, although I suspect not.

---

EDIT 1: compiled the app with `yarn build` and the behavior is pretty much the same as stated above. However, there are a few glitches. Recorded a short screencast demonstrating: https://www.loom.com/share/444e7f546ff1441fb6a8842157f399ef.

---

Any thoughts? 🤔

**Reply by CHUYKOPS** [reply]
I can say for our team that such an opportunity would be convenient. You can make a separate article and specify there a list of languages ​​that can be included in the code block. It's comfortable.
Changing the list of languages ​​can be set, for example, either in the admin panel. And if this is not convenient, then you can specify a list in the collection settings.

**Reply by tommoor** [reply]
Nice, lets put a PR in and we can go from there – I think the behavior you described should be easy enough to fix.

**Reply by CHUYKOPS** [reply]
Nice! Also will be great to do same thing with integration links.
